sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala

```diff
@@ -76,13 +76,17 @@ case class EnsureRequirements(
       case _ => false
     }.map(_._2)
 
-    // Special case: if all sides of the join are single partition
-    val allSinglePartition =
-      childrenIndexes.forall(children(_).outputPartitioning == SinglePartition)
+    // Special case: if all sides of the join are single partition and it's physical size less than
+    // or equal advisoryPartitionSizeInBytes.
+    val preferSinglePartition = childrenIndexes.forall { i =>
+      children(i).outputPartitioning == SinglePartition &&
+        children(i).logicalLink
+          .forall(_.stats.sizeInBytes <= conf.getConf(SQLConf.ADVISORY_PARTITION_SIZE_IN_BYTES))
```

Comment:
Looks OK, but I'm not totally sure whether it's good to reuse `SQLConf.ADVISORY_PARTITION_SIZE_IN_BYTES` here. It is for AQE and this code is used by both AQE and non-AQE?

Reply:
How about `SQLConf.FILES_MAX_PARTITION_BYTES`?

Reply:
Added a new config: spark.sql.maxSinglePartitionBytes